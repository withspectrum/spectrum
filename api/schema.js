// @flow
/**
 * The combined schema out of types and resolvers (queries, mutations and subscriptions)
 */
const {
  makeExecutableSchema,
  addSchemaLevelResolveFunction,
} = require('graphql-tools');
const debug = require('debug')('api:resolvers');
const logExecutions = require('graphql-log')({
  logger: debug,
});
const { merge } = require('lodash');
import UserError from './utils/UserError';

const scalars = require('./types/scalars');
const generalTypes = require('./types/general');

const Thread = require('./types/Thread');
const Channel = require('./types/Channel');
const ChannelSlackSettings = require('./types/ChannelSlackSettings');
const Community = require('./types/Community');
const CommunitySlackSettings = require('./types/CommunitySlackSettings');
const Message = require('./types/Message');
const Reaction = require('./types/Reaction');
const User = require('./types/User');
const DirectMessageThread = require('./types/DirectMessageThread');
const Notification = require('./types/Notification');
const Meta = require('./types/Meta');
const Invoice = require('./types/Invoice');
const Search = require('./types/Search');
const CommunityMember = require('./types/CommunityMember');
const ThreadParticipant = require('./types/ThreadParticipant');

const ThreadQueries = require('./queries/thread');
const channelQueries = require('./queries/channel');
const communityQueries = require('./queries/community');
const messageQueries = require('./queries/message');
const userQueries = require('./queries/user');
const reactionQueries = require('./queries/reaction');
const directMessageThreadQueries = require('./queries/directMessageThread');
const notificationQueries = require('./queries/notification');
const metaQueries = require('./queries/meta');
const searchQueries = require('./queries/search');
const communityMemberQueries = require('./queries/communityMember');
const communitySlackSettingsQueries = require('./queries/communitySlackSettings');
const channelSlackSettingsQueries = require('./queries/channelSlackSettings');

const messageMutations = require('./mutations/message');
const threadMutations = require('./mutations/thread');
const reactionMutations = require('./mutations/reaction');
const recurringPaymentMutations = require('./mutations/recurringPayment');
const communityMutations = require('./mutations/community');
const channelMutations = require('./mutations/channel');
const directMessageThreadMutations = require('./mutations/directMessageThread');
const notificationMutations = require('./mutations/notification');
const userMutations = require('./mutations/user');
const metaMutations = require('./mutations/meta');
const communityMemberMutations = require('./mutations/communityMember');

const messageSubscriptions = require('./subscriptions/message');
const notificationSubscriptions = require('./subscriptions/notification');
const directMessageThreadSubscriptions = require('./subscriptions/directMessageThread');
const threadSubscriptions = require('./subscriptions/thread');

const Root = /* GraphQL */ `
  # The dummy queries and mutations are necessary because
  # graphql-js cannot have empty root types and we only extend
  # these types later on
  # Ref: apollographql/graphql-tools#293
  type Query {
    dummy: String
  }

  type Mutation {
    dummy: String
  }

  type Subscription {
    dummy: String
  }

  schema {
    query: Query
    mutation: Mutation
    subscription: Subscription
  }
`;

const resolvers = merge(
  {},
  //queries
  scalars.resolvers,
  ThreadQueries,
  channelQueries,
  communityQueries,
  messageQueries,
  userQueries,
  directMessageThreadQueries,
  reactionQueries,
  notificationQueries,
  metaQueries,
  searchQueries,
  communityMemberQueries,
  communitySlackSettingsQueries,
  channelSlackSettingsQueries,
  // mutations
  messageMutations,
  threadMutations,
  directMessageThreadMutations,
  reactionMutations,
  recurringPaymentMutations,
  communityMutations,
  channelMutations,
  notificationMutations,
  userMutations,
  metaMutations,
  communityMemberMutations,
  // subscriptions
  messageSubscriptions,
  notificationSubscriptions,
  directMessageThreadSubscriptions,
  threadSubscriptions
);

if (process.env.NODE_ENV === 'development' && debug.enabled) {
  logExecutions(resolvers);
}

// Create the final GraphQL schema out of the type definitions
// and the resolvers
const schema = makeExecutableSchema({
  typeDefs: [
    scalars.typeDefs,
    generalTypes,
    Root,
    Community,
    CommunitySlackSettings,
    CommunityMember,
    Channel,
    ChannelSlackSettings,
    Thread,
    ThreadParticipant,
    Message,
    Reaction,
    User,
    DirectMessageThread,
    Notification,
    Meta,
    Invoice,
    Search,
  ],
  resolvers,
});

module.exports = schema;
