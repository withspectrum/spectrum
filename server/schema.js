//@flow
/**
 * The combined schema out of types and resolvers (queries, mutations and subscriptions)
 */
//$FlowFixMe
const { makeExecutableSchema } = require('graphql-tools');
//$FlowFixMe
const { merge } = require('lodash');
//$FlowFixMe
const { maskErrors } = require('graphql-errors');

const scalars = require('./types/scalars');
const generalTypes = require('./types/general');

const Story = require('./types/Story');
const Frequency = require('./types/Frequency');
const Community = require('./types/Community');
const Message = require('./types/Message');
const Reaction = require('./types/Reaction');
const User = require('./types/User');
const DirectMessageGroup = require('./types/DirectMessageGroup');
const Notification = require('./types/Notification');

const storyQueries = require('./queries/story');
const frequencyQueries = require('./queries/frequency');
const communityQueries = require('./queries/community');
const messageQueries = require('./queries/message');
const userQueries = require('./queries/user');
const reactionQueries = require('./queries/reaction');
const directMessageGroupQueries = require('./queries/directMessageGroup');
const notificationQueries = require('./queries/notification');

const messageMutations = require('./mutations/message');
const storyMutations = require('./mutations/story');
const reactionMutations = require('./mutations/reaction');
const communityMutations = require('./mutations/community');
const frequencyMutations = require('./mutations/frequency');
const directMessageGroupMutations = require('./mutations/directMessageGroup');
const notificationMutations = require('./mutations/notification');
const userMutations = require('./mutations/user');

const messageSubscriptions = require('./subscriptions/message');

const Root = /* GraphQL */ `
	# The dummy queries and mutations are necessary because
	# graphql-js cannot have empty root types and we only extend
	# these types later on
	# Ref: apollographql/graphql-tools#293
	type Query {
		dummy: String
	}

	type Mutation {
		dummy: String
	}

	type Subscription {
		dummy: String
	}

	schema {
		query: Query
		mutation: Mutation
		subscription: Subscription
	}
`;

// Create the final GraphQL schema out of the type definitions
// and the resolvers
const schema = makeExecutableSchema({
  typeDefs: [
    scalars.typeDefs,
    generalTypes,
    Root,
    Community,
    Frequency,
    Story,
    Message,
    Reaction,
    User,
    DirectMessageGroup,
    Notification,
  ],
  resolvers: merge(
    {},
    //queries
    scalars.resolvers,
    storyQueries,
    frequencyQueries,
    communityQueries,
    messageQueries,
    userQueries,
    directMessageGroupQueries,
    reactionQueries,
    notificationQueries,
    // mutations
    messageMutations,
    storyMutations,
    directMessageGroupMutations,
    reactionMutations,
    communityMutations,
    frequencyMutations,
    notificationMutations,
    userMutations,
    // subscriptions
    messageSubscriptions
  ),
});

if (process.env.NODE_ENV === 'production') {
  maskErrors(schema);
}

module.exports = schema;
