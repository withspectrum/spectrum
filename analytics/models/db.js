// @flow
/**
 * Database setup is done here
 */
const fs = require('fs');
const path = require('path');
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max: 500, // Maximum number of connections, default is 1000
  buffer: 5, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
};

const PRODUCTION_CONFIG = {
  password: process.env.COMPOSE_US_WEST_2_RETHINKDB_PASSWORD,
  user: 'admin',
  host: process.env.COMPOSE_US_WEST_2_RETHINKDB_URL,
  port: process.env.COMPOSE_US_WEST_2_RETHINKDB_PORT,
  ssl: {
    ca: fs.readFileSync(path.join(process.cwd(), 'cacert')),
  },
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

module.exports = { db: r };
