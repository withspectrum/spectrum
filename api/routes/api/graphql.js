// @flow
import { graphqlExpress } from 'graphql-server-express';
import depthLimit from 'graphql-depth-limit';
import costAnalysis from 'graphql-cost-analysis';
import { crunch } from 'graphql-crunch';
import UserError from '../../utils/UserError';
import createLoaders from '../../loaders/';
import createErrorFormatter from '../../utils/create-graphql-error-formatter';
import schema from '../../schema';

export default graphqlExpress(req => {
  const loaders = createLoaders();

  let currentUser = req.user && !req.user.bannedAt ? req.user : null;

  return {
    schema,
    formatError: createErrorFormatter(req),
    formatResponse: response => {
      if (
        req.query.crunch !== undefined &&
        response.data &&
        !response.data.__schema
      ) {
        response.data = crunch(response.data);
      }

      return response;
    },
    context: {
      loaders,
      user: currentUser,
    },
    validationRules: [
      depthLimit(10),
      costAnalysis({
        variables: req.body.variables,
        maximumCost: 750,
        defaultCost: 1,
        createError: (max, actual) => {
          const err = new UserError(
            `GraphQL query exceeds maximum complexity, please remove some nesting or fields and try again. (max: ${max}, actual: ${actual})`
          );
          return err;
        },
      }),
    ],
  };
});
