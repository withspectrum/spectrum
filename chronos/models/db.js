/**
 * Database setup is done here
 */
const fs = require('fs');
const path = require('path');
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  db: 'spectrum',
};

const PRODUCTION_CONFIG = {
  password: process.env.AWS_RETHINKDB_PASSWORD,
  host: process.env.AWS_RETHINKDB_URL,
  port: process.env.AWS_RETHINKDB_PORT,
};

const config = !IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

module.exports = { db: r };
