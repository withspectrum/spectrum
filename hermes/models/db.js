// @flow
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  db: 'spectrum',
  max: 100, // Maximum number of connections, default is 1000
  buffer: 10, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
};

const PRODUCTION_CONFIG = {
  password: process.env.AWS_RETHINKDB_PASSWORD,
  host: process.env.AWS_RETHINKDB_URL,
  port: process.env.AWS_RETHINKDB_PORT,
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

const r = require('rethinkdbdash')(config);

module.exports = { db: r };
