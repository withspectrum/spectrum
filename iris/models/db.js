// @flow
/**
 * Database setup is done here
 */
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max: 500, // Maximum number of connections, default is 1000
  buffer: 5, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
  pingInterval: 300, // Ping the connection every 5 minutes (300 seconds) to keep it alive and prevent rethinkdbdash#192
};

const PRODUCTION_CONFIG = {
  password: process.env.AWS_RETHINKDB_PASSWORD,
  host: process.env.AWS_RETHINKDB_URL,
  port: process.env.AWS_RETHINKDB_PORT,
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

if (process.env.NODE_ENV === 'development') {
  const fs = require('fs');
  const inspect = require('rethinkdb-inspector');
  const queries = [];
  inspect(r, {
    onQueryComplete: (query, { size, time }) => {
      if (query.indexOf('.changes') > -1) return;
      queries.push({ query, time, size });
      fs.writeFileSync(
        'queries-by-time.js',
        JSON.stringify(queries.sort((a, b) => b.time - a.time), null, 2)
      );
      fs.writeFileSync(
        'queries-by-response-size.js',
        JSON.stringify(queries.sort((a, b) => b.size - a.size), null, 2)
      );
    },
  });
}

module.exports = { db: r };
