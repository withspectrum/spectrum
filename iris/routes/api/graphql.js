// @flow
import { graphqlExpress } from 'graphql-server-express';
import depthLimit from 'graphql-depth-limit';
import costAnalysis from 'graphql-cost-analysis';
import Raven from 'shared/raven';
import UserError from '../../utils/UserError';
import createLoaders from '../../loaders/';

import createErrorFormatter from '../../utils/create-graphql-error-formatter';
import schema from '../../schema';

export default graphqlExpress(req => ({
  schema,
  formatError: createErrorFormatter(req),
  // Only add tracing information in production, otherwise Apollo Engine would send
  // performance data of development machines to the remote server which is not what we want
  tracing: process.env.NODE_ENV === 'production',
  context: {
    user: req.user,
    loaders: createLoaders(),
  },
  validationRules: [
    depthLimit(10),
    costAnalysis({
      variables: req.body.variables,
      maximumCost: 750,
      defaultCost: 1,
      createError: (max, actual) => {
        const err = new UserError(
          `GraphQL query exceeds maximum complexity, please remove some nesting or fields and try again. (max: ${max}, actual: ${actual})`
        );
        return err;
      },
    }),
  ],
}));
