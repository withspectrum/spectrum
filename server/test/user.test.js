import { graphql } from 'graphql';
import { db as mockTestDb, setup, teardown, data } from './db';
jest.mock('../models/db', () => ({
  db: mockTestDb,
}));

import schema from '../schema';

describe('queries', () => {
  beforeAll(() => setup(mockTestDb));
  afterAll(() => teardown(mockTestDb));

  it('should fetch a user', () => {
    const query = /* GraphQL */ `
			{
				user(id: "first-user") {
					uid
					createdAt
					lastSeen
					photoURL
					displayName
					username
					email
				}
			}
		`;

    expect.assertions(1);
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  it('should return null for a non-existant id', () => {
    const query = /* GraphQL */ `
			{
				user(id: "non-existant") {
					uid
				}
			}
		`;

    expect.assertions(1);
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  describe('everything', () => {
    it('should return the latest story', () => {
      const query = /* GraphQL */ `
  			{
  				user(id: "first-user") {
  					everything(first: 1) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      expect.assertions(1);
      return graphql(schema, query).then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    it('should paginate based on the after property', () => {
      const query = /* GraphQL */ `
  			{
  				user(id: "first-user") {
  					everything(first: 1, after: "c2Vjb25kLXN0b3J5") {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      expect.assertions(1);
      return graphql(schema, query).then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    it('should handle first being set to 0 correctly', () => {
      const noCursorQuery = /* GraphQL */ `
  			{
  				user(id: "first-user") {
  					everything(first: 0) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      const cursorQuery = /* GraphQL */ `
  			{
  				user(id: "first-user") {
  					everything(first: 0, after: "Zmlyc3Qtc3Rvcnk=") {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      expect.assertions(2);
      return graphql(schema, noCursorQuery)
        .then(result => {
          expect(result).toMatchSnapshot();
          return graphql(schema, cursorQuery);
        })
        .then(result => {
          expect(result).toMatchSnapshot();
        });
    });
  });

  it.skip('fetches a users communities');
  it.skip('fetches a users frequencies');
});
