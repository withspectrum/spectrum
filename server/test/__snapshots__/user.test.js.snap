exports[`queries everything should handle first being set to 0 correctly 1`] = `
Object {
  "data": Object {
    "user": Object {
      "everything": Object {
        "edges": Array [],
        "pageInfo": Object {
          "hasNextPage": true,
        },
      },
    },
  },
}
`;

exports[`queries everything should handle first being set to 0 correctly 2`] = `
Object {
  "data": Object {
    "user": Object {
      "everything": Object {
        "edges": Array [],
        "pageInfo": Object {
          "hasNextPage": false,
        },
      },
    },
  },
}
`;

exports[`queries everything should paginate based on the after property 1`] = `
Object {
  "data": Object {
    "user": Object {
      "everything": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "first-story",
            },
          },
        ],
        "pageInfo": Object {
          "hasNextPage": false,
        },
      },
    },
  },
}
`;

exports[`queries everything should return the latest story 1`] = `
Object {
  "data": Object {
    "user": Object {
      "everything": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "second-story",
            },
          },
        ],
        "pageInfo": Object {
          "hasNextPage": true,
        },
      },
    },
  },
}
`;

exports[`queries should fetch a user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "createdAt": "2017-01-01T23:00:00.000Z",
      "displayName": "First User",
      "email": "first.user@gmail.com",
      "lastSeen": "2017-02-01T23:00:00.000Z",
      "photoURL": "my-photo.jpg",
      "uid": "first-user",
      "username": "first",
    },
  },
}
`;

exports[`queries should return null for a non-existant id 1`] = `
Object {
  "data": Object {
    "user": null,
  },
}
`;
